Check for customer before checking for VAT exemption

pierrebuet opened the pull request:
In the actual form, the check done in the function is_taxable()  of a shipping method can only be done in the frontend, since WC()->customer must be defined.  Checking for a customer before checking the excemption would make the function also callable in the backend for different pre-calculation. If a customer is not defined, an exemption can not exists.

### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes # .

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a short summary of all changes on this Pull Request. This will appear in the changelog if accepted.

pierrebuet committed:
Check for customer before checking for VAT excemption

In the actual form, the check done in the function is_taxable()  of a shipping method can only be done in the frontend, since WC()->customer must be defined.  Checking for a customer before checking the excemption would make the function also callable in the backend for different pre-calculation. If a customer is not defined, an exemption can not exists.

claudiulodro reviewed (approved):
Makes sense. Looks good. Thanks for the contribution.

pierrebuet committed:
Sort checkout fields by field priority

pierrebuet commented:
Sorry for the big mess, but I just don't know how to pull a request when another one was still opened.
Now some background information for the second commit concerning the checkout fields:
Before https://github.com/woocommerce/woocommerce/commit/11da1898b83832057b01eec921c5e881c7cdc509#diff-27a95ce2f13e565129eca0979b69aedb,

the sorting of the checkout fields was done on client side on form loading, based on the priority of the fields. Changing the country would again trigger the reordering of the field done through address-i18n.js

Since this initial trigger for reordering doesn't happen anymore on form loading, we need to do the job address-i18n.js was doing on client side while loading but now on server side, reordering the checkout fields based on their priority attributes. Otherwise we get the array of fields in the default order with updated priorities through locale, but not in the order based on priority.

As an exemple for german, locale is changing the priority of the postcode field to have it displayed before the city.  The server is answering with an array of field where the priority of postcode was changed, but not put before the city field. This looks this way:
![before](https://user-images.githubusercontent.com/25102786/47724263-9df24a80-dc56-11e8-87ed-25e1ae89ef44.jpeg)
Doing on server side the reordering the client would do with java and sorting the fields based on their priority looks this way:
![after](https://user-images.githubusercontent.com/25102786/47724323-bd897300-dc56-11e8-85c1-30d3b373780d.jpeg)

claudiulodro commented:
@pierrebuet I believe a fix for the reordering issue was merged in https://github.com/woocommerce/woocommerce/pull/21733 (by adding the JS trigger back). Can you please revert the latest commit? Thanks! :)

kloon reviewed (changes requested):
Can you please remove the commit with the priority change in? I have a new PR as the changes you made in that commit is not working for all the filters you can use to alter the checkout fields. I will give you credit in the PR for the intial work you did, but took some reworking to get it 100% right. Thanks!

I think `git revert 7e8ada187e2c73ca3a87a6d41a54070051c43992` should do the trick for you.

pierrebuet committed:
Revert "Sort checkout fields by field priority"

This reverts commit 7e8ada187e2c73ca3a87a6d41a54070051c43992.

pierrebuet commented:
I hope I revert the commit correctly now. Sorry for all the trouble with the git stuff.

pierrebuet committed:
Correct misleading deprecated notice in WP 5.0

When editing a standard post in WP 5.0, the console print out a deprecated notice for edit_form_after_title. To overcome this, we should not hook and then test the post type, but only hook on the right screen (coupons).

pierrebuet commented:
The commit  "Correct misleading deprecated notice in WP 5.0" corrects issue #21962.
Unfortunately in WP 5.0, post.php checks - when using the block editor - all functions hooked to edit_form_after_title and emits a deprecated notice, no matter if the function output something or not. The only solution to get rid of the warning is to only hook the function when needed - in this case when editing a coupon in admin - instead of hooking on all pages and checking within the hooked function for the post type.

mikejolley commented:
I'm going to cherry pick the commit that fixes the customer object only.

mikejolley commented:
Done. Thanks!
